feat: move KeepKey's pin changing behavior to the wallet menu

NeOMakinG opened the pull request:
## Description
- Moved the `Pin` logic and markup to a reusable component and added some props needed to customize it for the menu
- Fixed a bug crashing the app inside error handling of the new pin action
- Added a new wallet state `isUpdatingPin` in order to prevent the regular modal to open when the reset pin action is launched from the menu
- Added a new wallet state `isLoading` so we can force a loading state when the device is loading for example
- I used polling to handle the passphrase sent event response, because the KeepKey is silent on success, see code comments for more information!

⚠️ There is a keepkey bug when cancelling / going back to previous route, the `cancel` method from hdwallet makes the keepkey crazy, I guess this is a firmware issue and will be fixed with https://github.com/shapeshift/web/issues/1613

<!-- Please describe your changes -->

## Notice

<!-- Before submitting a pull request, please make sure you have answered the following: -->

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #1414

## Risk

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

- The reset pin feature could be broken or buggy
- The enter pin feature could be broken or buggy in case I did an error while moving to a reusable component

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

- Connect your keepkey
- Open the wallet menu on the top-right and try to reset the pin of the keepkey
- Complete the whole pin reset process, everything should work as expected until the end
- Try to also cancel, go back to the previous menu screen
- You'll have more information in the issue, mockups and as a video in the screenshots part

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

- Connect your keepkey
- Open the wallet menu on the top-right and try to reset the pin of the keepkey
- Complete the whole pin reset process, everything should work as expected until the end
- Try to also cancel, go back to the previous menu screen
- You'll have more information in the issue, mockups and as a video in the screenshots part

### Operations

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

- Connect your keepkey
- Open the wallet menu on the top-right and try to reset the pin of the keepkey
- Complete the whole pin reset process, everything should work as expected until the end
- Try to also cancel, go back to the previous menu screen
- You'll have more information in the issue, mockups and as a video in the screenshots part

## Screenshots (if applicable)
https://www.awesomescreenshot.com/video/11006360?key=80a2f395eaa20c19c602b2d95329e46f

0xApotheosis commented on src/context/WalletProvider/KeepKey/hooks/useKeepKeyEventHandler.ts:
```diff
@@ -114,6 +122,10 @@ export const useKeepKeyEventHandler = (
         case MessageType.PINMATRIXACK:
           if (modal) dispatch({ type: WalletActions.SET_WALLET_MODAL, payload: false })
           break
+        // A bit tricky but if we receive this event, the device is obviously not waiting for interaction
```
This assertion is not true. We can absolutely receive a `ETHEREUMGETADDRESS` whilst waiting for a device interaction. 

Indeed, this seems to be causing a bug where the UI will think it's no longer awaiting a device interaction whilst the KeepKey is awaiting for PIN change confirmation.

0xApotheosis reviewed (changes requested):
I haven't yet done a deep review of the implementation, but I've noticed a few things on first pass:

1. Setting `awaitingDeviceInteraction` to `false` on `ETHEREUMGETADDRESS` events is causing bugs. The assertion that it means we aren't waiting for a device interaction is not true.
2. We should add a loading state to the UI once whilst we are verifying the PIN, before we start the new PIN entry process. At the moment it just hangs. We'll likely want to stay in a loading state until we get a `PIN_REQUEST` event back from the KeepKey. 
3. The state does not get reset correctly when going back/cancelling a PIN change request whilst entering a new PIN. It leaves the device/app in a broken state
4. The PIN entry UI is a bit rough, and should probably match the aesthetic of the existing PIN entry modal (dots on buttons-like elements, rather than just dots).

KeepKey is a tricky beast. We need to be really careful that we (a) always send cancel events to the KeepKey whenever the user backs out of something (intentionally, or not), and (b) ensure that the `DeviceState` stays in sync with the UI (this one is particularly tricky, as if we miss a salient event we get out of sync).

NeOMakinG commented:
@0xApotheosis 

As we discussed, I addressed all your feedback and finally used polling to handle the keepkey passphrase response event.

Unfortunately, I can't fix the cancel problem, it already exists in the current reset pin implementation and is probably a firmware thing that you already added to another issue (see the description of the PR).

Thanks for the great feedbacks and tests 👍

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
+    setDeviceState,
   } = useWallet()
   const toast = useToast()
+  const pinButtonsBackground = useColorModeValue('gray.200', 'gray.600')
```
```suggestion
  const pinButtonBackground = useColorModeValue('gray.200', 'gray.600')
```

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
   } = useWallet()
   const toast = useToast()
+  const pinButtonsBackground = useColorModeValue('gray.200', 'gray.600')
+  const pinButtonsBackgroundHover = useColorModeValue('gray.100', 'gray.500')
```
```suggestion
  const pinButtonBackgroundHover = useColorModeValue('gray.100', 'gray.500')
```

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
       })
     })
 
+    setDeviceState({
+      isUpdatingPin: false,
+    })
```
For correctness, this should be in a `finally` block of `changePin()`.

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
+      })
+    })
+
+    setDeviceState({
```
For correctness, this should be in the `finally` block of `cancel()`.

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
   }
   const setting = 'PIN'
 
+  const shouldDisplayPinView = isUpdatingPin && !awaitingDeviceInteraction
```
Slight naming improvement for clarity.

```suggestion
  const shouldDisplayEntryPinView = isUpdatingPin && !awaitingDeviceInteraction
```

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
 
+  const shouldDisplayPinView = isUpdatingPin && !awaitingDeviceInteraction
+
   const renderPinState: JSX.Element = (() => {
```
I don't know what I was thinking when I named this initially, but because it's an IIFE, it's rendered statically, and so should actually be `renderedPinState`.

```suggestion
  const renderedPinState: JSX.Element = (() => {
```

0xApotheosis commented on src/context/WalletProvider/KeepKey/KeepKeyTypes.ts:
```diff
@@ -93,6 +93,8 @@ export enum MessageType {
   RESETDEVICE = 14,
   SIGNTX = 15,
   FEATURES = 17,
+  PINREQUEST = 'PIN_REQUEST',
+  PINCHANGED = 'PIN changed',
```
These are _messages_, not _message types_. We need to either use the numerical code with these message types, or put these into a new `Message` enum and use that.

The former is preferred.

0xApotheosis commented on src/components/Layout/Header/NavBar/KeepKey/ChangePin.tsx:
```diff
@@ -1,50 +1,139 @@
+  const handleHeaderBackClick = async () => {
+    await handleCancel()
+
+    handleBackClick()
```
We're missing an `await`.

```suggestion
    await handleBackClick()
```

0xApotheosis commented on src/components/Layout/Header/NavBar/SubmenuHeader.tsx:
```diff
@@ -8,9 +8,10 @@ type ExpandedMenuItemProps = {
 
-export const SubmenuHeader = ({ title, description }: ExpandedMenuItemProps) => {
+export const SubmenuHeader = ({ title, description, onBackClick }: ExpandedMenuItemProps) => {
   const { handleBackClick } = useMenuRoutes()
```
Now that we effectively have two `handleBackClick`s, rename this one to `handleBackClickDefault` on destructure.

```suggestion
  const { handleBackClick: handleBackClickDefault } = useMenuRoutes()
```

0xApotheosis commented on src/components/Layout/Header/NavBar/SubmenuHeader.tsx:
```diff
@@ -8,9 +8,10 @@ type ExpandedMenuItemProps = {
 }
 
-export const SubmenuHeader = ({ title, description }: ExpandedMenuItemProps) => {
+export const SubmenuHeader = ({ title, description, onBackClick }: ExpandedMenuItemProps) => {
```
```suggestion
export const SubmenuHeader = ({ title, description, onBackClick: handleBackClick }: ExpandedMenuItemProps) => {
```

0xApotheosis commented on src/components/Layout/Header/NavBar/SubmenuHeader.tsx:
```diff
@@ -21,7 +22,7 @@ export const SubmenuHeader = ({ title, description }: ExpandedMenuItemProps) =>
           isRound
           size='sm'
-          onClick={handleBackClick}
+          onClick={onBackClick ?? handleBackClick}
```
Once the above changes are made:

```suggestion
          onClick={handleBackClick ?? handleBackClickDefault}
```

0xApotheosis commented on src/context/WalletProvider/KeepKey/hooks/useKeepKeyEventHandler.ts:
```diff
@@ -91,17 +99,47 @@ export const useKeepKeyEventHandler = (
+
+          // KeepKey doesn't send a successfull response on passphrase sent request
+          // So, we need to poll to keep the DeviceState synced
+          poll({
```
This is quite the hack, but it works, and seems safe 👌

0xApotheosis commented on src/context/WalletProvider/WalletProvider.tsx:
```diff
@@ -279,6 +290,16 @@ const reducer = (state: InitialState, action: ActionTypes) => {
     case WalletActions.RESET_STATE:
       const resetProperties = omit(initialState, ['keyring', 'adapters', 'modal', 'deviceId'])
       return { ...state, ...resetProperties }
+    case WalletActions.RESET_LAST_DEVICE_INTERACTION_STATE: {
+      const { deviceState } = state
+      return {
+        ...state,
+        deviceState: {
+          ...deviceState,
+          lastDeviceInteractionStatus: undefined,
+        },
+      }
+    }
```
Do we really need this?

Can't we just use `SET_DEVICE_STATE` with a payload of `{ lastDeviceInteractionStatus: undefined }`?

NeOMakinG replied:
You can't, because `SET_DEVICE_STATE` use the current state value as default value in case the state is not provided in the payload, it permits to avoid to add every states when you use it

If you do this: `setDeviceState({ lastDeviceInteractionStatus: undefined })`, it will keep the current value of `lastDeviceInteractionStatus` which could be `true`, `false`, `anything else`...

0xApotheosis replied:
Right you are, `SET_DEVICE_STATE` doesn't currently support explicitly setting `undefined` values, which is a design flaw.

This is not a concern for this PR, but something we should fix.

I added a TODO in https://github.com/shapeshift/web/pull/2589/commits/3e90f5ca2601ab71c136907bbcaa68664f141106

0xApotheosis commented on src/context/WalletProvider/WalletProvider.tsx:
```diff
@@ -68,6 +68,8 @@ export type DeviceState = {
   recoveryCharacterIndex: number | undefined
   recoveryWordIndex: number | undefined
+  isUpdatingPin: boolean | undefined
+  isLoading: boolean | undefined
```
What does this represent? It feels closer to a property of the UI than the device.

NeOMakinG replied:
The KeepKey isn't telling the application that it's currently loading, so it should probably be used carefully but this is a state that I added in order to tell the UI that the keepkey is loading at this time

When you enter the new pin, for example, the keepkey enter a loading state, the UI needs to know it in order to inform the user that it's currently loading...

0xApotheosis commented on src/context/WalletProvider/KeepKey/components/PinModal.tsx:
```diff
@@ -0,0 +1,35 @@
+        <Text translation={`walletProvider.keepKey.${translationType}.header`} />
+      </ModalHeader>
+      <ModalBody>
+        <KeepKeyPin translationType={translationType} />
```
This is a nice decoupling of concerns, nice.

0xApotheosis commented on src/context/WalletProvider/actions.ts:
```diff
@@ -16,13 +16,15 @@ export enum WalletActions {
   SET_IS_LOCKED = 'SET_IS_LOCKED',
   SET_WALLET_MODAL = 'SET_WALLET_MODAL',
   RESET_STATE = 'RESET_STATE',
+  RESET_LAST_DEVICE_INTERACTION_STATE = 'RESET_LAST_DEVICE_INTERACTION_STATE',
```
See comment above, I'm not convinced we need this.

0xApotheosis reviewed (changes requested):
Looking much better this pass @NeOMakinG.

I've reviewed in depth this time around, and added some commentary on architecture.

Loading states are looking good, though we could still use one between setting a new pin, and confirming it - at the moment there is a flash as the old component disappears, and the new one loads.

0xApotheosis reviewed (approved):
Solid PR @NeOMakinG 👌 

Approving with the note that there are still some strange behaviours when cancelling an action, but it's outside the scope of this PR.

This works well with https://github.com/shapeshift/web/pull/2689, too, with keyboard entry and "enter" event support.

🍰
